### PR TITLE
[Feature] Add possibility to override whole user party

### PR DESCRIPTION
**Pull request by hayuna**
## What are the changes?
Add possibility to change whole user team

## Why am I doing these changes?
Using this solution, we have more flexibility to use specific pokemon.

## What did change?
Replace STARTER_SPECIES_OVERRIDE from number|species to array of numbers|species
Replace STARTER_FORM_OVERRIDE from number to array of numbers
Replace ABILITY_OVERRIDE from Ability to array of Abilities

### Screenshots/Videos
Screenshot 2024-05-31 at 13 35 30

Screenshot 2024-05-31 at 19 01 55

## How to test the changes?
```javascript
export const STARTER_OVERRIDE: StarterOverride[] = [
  {
    species: Species.VIVILLON,
    form: 0,
    ability: Abilities.ADAPTABILITY,
    passiveAbility: Abilities.GALE_WINGS,
    status: StatusEffect.BURN,
    gender: Gender.FEMALE,
    moveset: [Moves.ABSORB, Moves.ACCELEROCK, Moves.ACID, Moves.ACID_ARMOR],
    shiny: false,
    shinyVariant: 0,
  },
  {
    species: Species.VIVILLON,
    form: 1,
    ability: Abilities.BAD_DREAMS,
    passiveAbility: Abilities.HADRON_ENGINE,
    status: StatusEffect.FAINT,
    gender: Gender.FEMALE,
    moveset: [Moves.BABY_DOLL_EYES, Moves.BELLY_DRUM, Moves.BLIZZARD, Moves.BODY_SLAM],
    shiny: false,
    shinyVariant: 1,
  },
  {
    species: Species.VIVILLON,
    form: 2,
    ability: Abilities.CHEEK_POUCH,
    passiveAbility: Abilities.ICE_BODY,
    status: StatusEffect.FREEZE,
    gender: Gender.GENDERLESS,
    moveset: [Moves.CALM_MIND, Moves.CAMOUFLAGE, Moves.CAPTIVATE, Moves.CONFUSION],
    shiny: false,
    shinyVariant: 2,
  },
  {
    species: Species.VIVILLON,
    form: 3,
    ability: Abilities.DAMP,
    passiveAbility: Abilities.JUSTIFIED,
    status: StatusEffect.NONE,
    gender: Gender.GENDERLESS,
    moveset: [Moves.DARKEST_LARIAT, Moves.DARK_PULSE, Moves.DARK_VOID, Moves.DREAM_EATER],
    shiny: true,
    shinyVariant: 0,
  },
  {
    species: Species.SHAYMIN,
    form: 0,
    ability: Abilities.EARLY_BIRD,
    passiveAbility: Abilities.KEEN_EYE,
    status: StatusEffect.PARALYSIS,
    gender: Gender.MALE,
    moveset: [Moves.EARTHQUAKE, Moves.EARTH_POWER, Moves.ENERGY_BALL, Moves.ECHOED_VOICE],
    shiny: true,
    shinyVariant: 1,
  },
  {
    species: Species.SHAYMIN,
    form: 1,
    ability: Abilities.FAIRY_AURA,
    passiveAbility: Abilities.LEAF_GUARD,
    status: StatusEffect.POISON,
    gender: Gender.MALE,
    moveset: [Moves.FACADE, Moves.FAIRY_LOCK, Moves.FAIRY_WIND, Moves.FAKE_TEARS],
    shiny: true,
    shinyVariant: 2,
  }
]
```

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [ ] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Comment by CodeTappert** [comment]
Well if this is done then you should not have to go to the select screen on "new game".

Also can you add that we can select the form for each of them

**Comment by hayuna** [comment]
@CodeTappert Need to deep dive into the code. Its not working the same for overriding ability, moveset etc. Will keep this PR as it is and think about overriding rest of attributes in the next pull requests

File: src/overrides.ts [hunk 1]
```diff
@@ -62,6 +62,11 @@ export const STARTING_LEVEL_OVERRIDE: integer = 0;
  * @example SPECIES_OVERRIDE = Species.Bulbasaur;
  */
 export const STARTER_SPECIES_OVERRIDE: Species | integer = 0;
+export const STARTER_SPECIES_OVERRIDE2: Species | integer = 0;
+export const STARTER_SPECIES_OVERRIDE3: Species | integer = 0;
+export const STARTER_SPECIES_OVERRIDE4: Species | integer = 0;
+export const STARTER_SPECIES_OVERRIDE5: Species | integer = 0;
+export const STARTER_SPECIES_OVERRIDE6: Species | integer = 0;
```
Comment (Tempo-anon) [inline src/overrides.ts:69]: With so many, would this make sense to make STARTER_SPECIES_OVERRIDE an array?

File: src/phases.ts [hunk 2]
```diff
@@ -529,8 +529,8 @@ export class SelectStarterPhase extends Phase {
         const party = this.scene.getParty();
         const loadPokemonAssets: Promise<void>[] = [];
         starters.forEach((starter: Starter, i: integer) => {
-          if (!i && Overrides.STARTER_SPECIES_OVERRIDE) {
-            starter.species = getPokemonSpecies(Overrides.STARTER_SPECIES_OVERRIDE as Species);
+          if (Overrides[`STARTER_SPECIES_OVERRIDE${i === 0 ? "" : i+1 }`]) {
```
Comment (Tempo-anon) [inline src/phases.ts:532]: 🤔

Comment (hayuna) [reply src/phases.ts:532]: Thats why first constants has no number in it. Using array, it could be better signed

**Comment by hayuna** [comment]
@CodeTappert Added possibility to override forms

File: src/overrides.ts [hunk 3]
```diff
@@ -62,7 +60,9 @@ export const STARTING_LEVEL_OVERRIDE: integer = 0;
  * default is 0 to not override
  * @example SPECIES_OVERRIDE = Species.Bulbasaur;
  */
-export const STARTER_SPECIES_OVERRIDE: Species | integer = 0;
+export const STARTER_SPECIES_OVERRIDE: Species[] | integer[] = [0, 0, 0, 0, 0, 0];
+// forms can be found in pokemon-species.ts
+export const STARTER_FORM_OVERRIDE: integer[] = [0, 0, 0, 0, 0, 0];
 export const ABILITY_OVERRIDE: Abilities = Abilities.NONE;
```
Comment (Tempo-anon) [inline src/overrides.ts:66]: Should all of these be arrays too?

Comment (ben-lear) [reply src/overrides.ts:66]: At this point, wouldn’t it make more sense to bundle all of these overrides together into one array of start object overrides. That way we don’t need to be juggling 6 different arrays

Comment (hayuna) [reply src/overrides.ts:66]: @ben-lear If I will find out how to override 6 pokemons Abilities, then will start thinking or merging these properties into one object 😄

**Comment by hayuna** [comment]
We can now override abilities for whole party

**Comment by hayuna** [comment]
- [ ] Species
- [ ] Forms
- [ ] Abilities
- [ ] Passive Abilities
- [ ] Statuses
- [ ] Genders
- [ ] Movesets
- [ ] Shinies
- [ ] Shiny Variants

**Review by Tempo-anon — APPROVED** [review]

**Pull request merged**